### ISSUE TITLE
changing delay time in audio echo‏

### ISSUE THREAD

**Issue by alexd555** [googlesamples/android-ndk#515]
Hi 

I tried to add seekBar for changing delay on RealTime but it doesn't work (no sound or app crash)
this my code(based on https://github.com/googlesamples/android-ndk/tree/master/audio-echo):

Could you help ?

Thanks,
Alex

```
int delay_factor=1;

@Override
protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    controlButton = (Button)findViewById((R.id.capture_control_button));
    statusView = (TextView)findViewById(R.id.statusView);
    toggleFilter = (Button)findViewById((R.id.toggle_filter_button));
    queryNativeAudioParameters();
    final SeekBar delayBar = (SeekBar)findViewById(R.id.delay_factor);
    delayBar.setOnSeekBarChangeListener(new OnSeekBarChangeListener() {
        @Override
        public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
            delay_factor=i+1;
            //  statusView.setText(""+Integer.parseInt(nativeSampleBufSize)*delay_factor);
            nativeSampleBufSize=""+Integer.parseInt(nativeSampleBufSize_base)*delay_factor;
            change_delay(delay_factor);
        }

        @Override
        public void onStartTrackingTouch(SeekBar seekBar) {

        }

        @Override
        public void onStopTrackingTouch(SeekBar seekBar) {

        }
    });
    // initialize native audio system
    updateNativeAudioUI();
    if (supportRecording) {
        createSLEngine(Integer.parseInt(nativeSampleRate), Integer.parseInt(nativeSampleBufSize));
    }
}

void change_delay(int delay_factor)
{
    if(isPlaying && supportRecording)
    {

        stopPlay();
        updateNativeAudioUI();
        deleteAudioRecorder();
        deleteSLBufferQueueAudioPlayer();
        deleteSLEngine();
        createSLEngine(Integer.parseInt(nativeSampleRate), Integer.parseInt(nativeSampleBufSize)*delay_factor);
        createSLBufferQueueAudioPlayer();
        createAudioRecorder();
        startPlay();
    }
    else if(supportRecording)
    {
       if (isPlaying) {
            stopPlay();
       }
       deleteSLEngine();
       createSLEngine(Integer.parseInt(nativeSampleRate), Integer.parseInt(nativeSampleBufSize)*delay_factor);
    }
}
```

**Comment by ggfan** [comment]
could you elaborate what the delay means? this sample there is no delay: coming out from recording directly feed into the playback; I could only guess that you want to add more latency at the very beginning: buffering the recorded audio, hold them until your required latency time satisfied, then begin to feed the buffered audio to the playback. is that correct?

If that is the case, you will need to allocate enough audio buffers to hold recorded audio samples for your whole delay time. 16 small buffer is not enough. May you check your code to see if it is related to this?  ( either play silent audio, or start playback later.  I would allocate enough buffers with silent audio in it, queue them up to the playback queue and start play and record at the same time. How many buffers to queue up ahead is a function of your required delay time).  Dynamically adjust latency time means restart record/playback engine.

**Comment by alexd555** [comment]
Yes. I want to adjust dynamically latency time.
What do you mean "restart" ? delete and create ?
I tried to do it on change_delay function but it dpesn't work.

What do you mean "16 small buffer is not enough" ?
for example, if I multiply buffer size on create then it works well :
createSLEngine(Integer.parseInt(nativeSampleRate), 10*Integer.parseInt(nativeSampleBufSize));

Could you help ?
@ggfan 

Thanks,
Alex

**Comment by ggfan** [comment]
I thank that should work: the delay is added with larger buffer size, then create the bufferSize ( as you do now ) with a function to your delay requirement:
```
 (delay_time_required_in_seconds / (nativeSampleBufSize / nativeSampleRate) ) * nativeSampleBufSize
 or simply put:
 delay_time_required_in_seconds * nativeSampleRate
```
that is your second parameters to createSLEngine(). you could not set a value smaller than the nativeSampleBufSize though.

the other way is to let C/C++ code to do it, I think your way is simpler.

**Comment by alexd555** [comment]
I don't understand your answer.
Why not ?        
createSLEngine(Integer.parseInt(nativeSampleRate), Integer.parseInt(nativeSampleBufSize)*delay_factor);

Thanks,
Alex 

@ggfan

**Comment by ggfan** [comment]
let me see what is your delay_factor definition

**Comment by alexd555** [comment]
it is letancy.
I only want to change it dynamically.
My issue is that I lost a signal.
Thanks,
Alex
@ggfan

**Comment by ggfan** [comment]
when audio is already echoing?

**Comment by alexd555** [comment]
When I start to change seekBar.
I think that it is becase of         deleteSLEngine()/Create .
I do someting wrong
@ggfan

**Comment by ggfan** [comment]
by doing the following on Pixel XL, it seems to be ok:
- removing create SL engine from onCreate()
- change the way onDestroy deleting SL engine to delete only when it is echoing
```
            stopPlay();  //this must include stopRecording()
            updateNativeAudioUI();
            deleteAudioRecorder();
            deleteSLBufferQueueAudioPlayer();
            //hacking...
            deleteSLEngine();
```
basically do the stop when it is echoing and user to close the app
- create and delete SL engine when as user clicks Start/Stop
```
    private void startEcho() {
        if(!supportRecording){
            return;
        }
        if (!isPlaying) {
            // hacking...
            createSLEngine(Integer.parseInt(nativeSampleRate), 10 * Integer.parseInt(nativeSampleBufSize));
            // hacking...
            if(!createSLBufferQueueAudioPlayer()) {
                statusView.setText(getString(R.string.error_player));
                return;
            }
            if(!createAudioRecorder()) {
                deleteSLBufferQueueAudioPlayer();
                statusView.setText(getString(R.string.error_recorder));
                return;
            }
            startPlay();   // this must include startRecording()
            statusView.setText(getString(R.string.status_echoing));
        } else {
            stopPlay();  //this must include stopRecording()
            updateNativeAudioUI();
            deleteAudioRecorder();
            deleteSLBufferQueueAudioPlayer();
            //hacking...
            deleteSLEngine();
            //hacking done
        }
        isPlaying = !isPlaying;
        controlButton.setText(getString((isPlaying == true) ?
                R.string.StopEcho: R.string.StartEcho));
    }
```
This works fine. Then in your case it should be ok, but your device may makes the difference? 

In your case, when user changes the factor, and when you get the user's delay_factor, you might do:
- do the similar change as the above
- if it is not playing (isPlaying is false), do nothing except remember it into the delay_factor; when user touches the start, it will automatically used
- if it is echoing ( isPlaying is true), just simply call startEcho(), if your startEcho() is doing the similar thing as the above code.

Once you get it working, you will have to put this create/delay etc into a different thread as create/delay takes time, your UI thread may not be happy. That does add complicity into your java code ( which I could only learn from you :-)), but it have to be done. For now you could try to get the functionality to work and ignore the warning.

**Comment by ggfan** [comment]
If you use your code as is, it might be closer as(did not test by myself):
```
// do not need to pass the delay_factor as it is a member function
void change_delay()
{
    if(!supportRecording) {
        return;
    }
    if(!isPlaying) {
        deleteSLEngine();
        createSLEngine(Integer.parseInt(nativeSampleRate), Integer.parseInt(nativeSampleBufSize)*delay_factor);
        return;
   }
      // in the middle of echo, do a full stop, delete SLEngine, create SLEngine, start it again.
    startEcho();  //this should stop it
    deleteSLEngine();
    createSLEngine(Integer.parseInt(nativeSampleRate), Integer.parseInt(nativeSampleBufSize)*delay_factor);
    startEcho(); // this should start it
    }
```

**Comment by ggfan** [comment]
do not worry about the thread things from Java side yet. Let's get your functionality to work

**Comment by alexd555** [comment]
Thanks
I will try and update you.

**Comment by ggfan** [comment]
For thread:    
Creating OpenSL Engine and deleting Engine takes time,  right now it is done on UI thread; UI thread does not allow long execution work ( your UI will expose as slow to your users ); at minimum you will get warning from logcat.  The sample only does it once at app startup and tearing down; but your app will constantly be doing it,  so you would get a warning in logcat for every engine creation.
The recommendation is to do those long execution tasks off UI thread: you create another thread and perform the creation/deletion in that thread.

After that you also need to have mutex mechanism to make sure your UI thread and working thread are nicely synchronized: what happen when the working thread is deleting/creating engine, user touches start/stop button?

Handling this action from Java side might be easier: you could refer to nn_sample in this repo for some ideas. yours will be different.  If you do this from cpp thread, then you will move those engine creation/deletion, player/recorder creation action bundles to cpp side, and when the required action is done, your cpp side will need to callback to Java side to notify UI thread that it is ready... It would be more work.

[one note that not relating to you: your usage is not what this sample tries to show as you app may not be working in low latency mode -- the buffer size is not the minimum buffer size. If your app plays silent audio for your required delay time, but keeps the buffer to be minimum size, then your app will be guaranteed working in "fast-mode". Anyhow it is not relevant to you as you wants more latency]

**Comment by alexd555** [comment]
Hi,
I'm not sure if I understand your note. But I'm aware that increasing of buffer size will increase latency. I want it.

 What do you mean "It would be more work" ? How much better ?
The ndk side is very complex for me. I prefer to do it on java.

Thanks,
Alex
@ggfan

**Comment by ggfan** [comment]
create a way that puts createSLEngine() related functions in its own thread when it is echoing, like putting startEcho() to be executed in a new working thread. 

If you have another start/stop button control, make sure synchronize both threads. That should do the work.
Note if you have a start/stop button + a slideBar, your function  change_delay() might need to check if currently is echoing -- if echo not started yet ( engine not created ), you do not need to call start_echo(). That way you leave start/stop to button control.  just my thoughts.

**Comment by alexd555** [comment]
Why can't I run twice startEcho ? This is more simpler.
Thanks,
Alex
@ggfan

**Comment by ggfan** [comment]
while that works, creating Engine takes time,  you probably want to minimize it. Your UI reaction might be flow to user ( not totally sure how much ).  either way works, I would do minimum work to avoid other potential errors I do not know.

**Comment by ggfan** [comment]
hard to guess: seems ok.  Android studio might point to certain line of code that causing the crash?

**Comment by ggfan** [comment]
there is lot of the off-line discussion concerning this. closing this one from here.
